light/client.go

```diff
@@ -737,43 +727,79 @@ func (c *Client) bisection(
 			}
 			depth++
 
-		case ErrInvalidHeader:
-			c.logger.Error("primary sent invalid header -> replacing", "err", err)
-			replaceErr := c.replacePrimaryProvider()
-			if replaceErr != nil {
-				c.logger.Error("Can't replace primary", "err", replaceErr)
-				// return original error
-				return fmt.Errorf("verify non adjacent from #%d to #%d failed: %w",
-					trustedHeader.Height, headerCache[depth].sh.Height, err)
-			}
-			newProviderHeader, newProviderVals, err := c.fetchHeaderAndValsAtHeight(newHeader.Height)
-			if err != nil {
-				return err
-			}
-			if !bytes.Equal(newProviderHeader.Hash(), newHeader.Hash()) || !bytes.Equal(newProviderVals.Hash(), newVals.Hash()) {
-				err := fmt.Errorf("replacement provider has a different header: %X and/or vals: %X at height: %d"+
-					"to the one being verified", newProviderHeader.Hash(), newProviderVals.Hash(), newHeader.Height)
-				return fmt.Errorf("verify non adjacent from #%d to #%d failed: %w",
-					trustedHeader.Height, headerCache[depth].sh.Height, err)
-			}
-			// attempt to verify the header again
-			return c.bisection(initiallyTrustedHeader, initiallyTrustedVals, newHeader, newVals, now)
-
 		default:
 			return fmt.Errorf("verify non adjacent from #%d to #%d failed: %w",
 				trustedHeader.Height, headerCache[depth].sh.Height, err)
 		}
 	}
 }
 
+// bisectionAgainstPrimary does bisection plus it compares new header with
+// witnesses and replaces primary if it does not respond after
+// MaxRetryAttempts.
+func (c *Client) bisectionAgainstPrimary(
+	initiallyTrustedHeader *types.SignedHeader,
+	initiallyTrustedVals *types.ValidatorSet,
+	newHeader *types.SignedHeader,
+	newVals *types.ValidatorSet,
+	now time.Time) error {
+
+	err := c.bisection(c.primary, initiallyTrustedHeader, initiallyTrustedVals, newHeader, newVals, now)
+
+	switch errors.Unwrap(err).(type) {
+	case ErrInvalidHeader:
+		c.logger.Error("primary sent invalid header -> replacing", "err", err)
+		replaceErr := c.replacePrimaryProvider()
+		if replaceErr != nil {
+			c.logger.Error("Can't replace primary", "err", replaceErr)
+			// return original error
+			return err
+		}
+
+		replacementHeader, replacementVals, fErr := c.signedHeaderAndValSetFromPrimary(newHeader.Height)
+		if fErr != nil {
+			c.logger.Error("Can't fetch header/vals from primary", "err", fErr)
+			// return original error
+			return err
+		}
+
+		if !bytes.Equal(replacementHeader.Hash(), newHeader.Hash()) ||
+			!bytes.Equal(replacementVals.Hash(), newVals.Hash()) {
+			c.logger.Error("Replacement provider has a different header/vals",
+				"newHash", newHeader.Hash(),
+				"newVals", newVals.Hash(),
+				"replHash", replacementHeader.Hash(),
+				"replVals", replacementVals.Hash())
+			// return original error
+			return err
+		}
+
+		// attempt to verify the header again
+		return c.bisection(c.primary, initiallyTrustedHeader, initiallyTrustedVals, replacementHeader, replacementVals, now)
```

Comment:
I think this needs to be recursive i.e. change to `c.bisectionAgainstPrimary` else if there is a replacement of providers, the result of the verification with this new provider will be directly returned and won't be compared with the witnesses (also as a slightly minor case, we won't be able to replace the primary more than once per bisection)

Reply:
good catch!